tests/e2e-inst-signatures/e2e-proctree_data_source.go

```diff
@@ -0,0 +1,304 @@
+package main
+
+import (
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/aquasecurity/tracee/pkg/utils"
+	"github.com/aquasecurity/tracee/signatures/helpers"
+	"github.com/aquasecurity/tracee/types/datasource"
+	"github.com/aquasecurity/tracee/types/detect"
+	"github.com/aquasecurity/tracee/types/protocol"
+	"github.com/aquasecurity/tracee/types/trace"
+)
+
+const (
+	bashPath  = "/usr/bin/bash"
+	sleepPath = "/usr/bin/sleep"
+	lsPath    = "/usr/bin/ls"
+)
+
+type e2eProcessTreeDataSource struct {
+	cb            detect.SignatureHandler
+	processTreeDS detect.DataSource
+}
+
+func (sig *e2eProcessTreeDataSource) Init(ctx detect.SignatureContext) error {
+	sig.cb = ctx.Callback
+	processTreeDataSource, ok := ctx.GetDataSource("tracee", "process_tree")
+	if !ok {
+		return fmt.Errorf("process tree data source not registered")
+	}
+	if processTreeDataSource.Version() > 1 {
+		return fmt.Errorf("process tree data source version not supported, please update this signature")
+	}
+	sig.processTreeDS = processTreeDataSource
+	return nil
+}
+
+func (sig *e2eProcessTreeDataSource) GetMetadata() (detect.SignatureMetadata, error) {
+	return detect.SignatureMetadata{
+		ID:          "PROCESS_TREE_DATA_SOURCE",
+		EventName:   "PROCESS_TREE_DATA_SOURCE",
+		Version:     "0.1.0",
+		Name:        "Process Tree Data Source Test",
+		Description: "Instrumentation events E2E Tests: Process Tree Data Source Test",
+		Tags:        []string{"e2e", "instrumentation"},
+	}, nil
+}
+
+func (sig *e2eProcessTreeDataSource) GetSelectedEvents() ([]detect.SignatureEventSelector, error) {
+	return []detect.SignatureEventSelector{
+		{Source: "tracee", Name: "sched_process_exec", Origin: "*"},
+	}, nil
+}
+
+func (sig *e2eProcessTreeDataSource) OnEvent(event protocol.Event) error {
+	eventObj, ok := event.Payload.(trace.Event)
+	if !ok {
+		return fmt.Errorf("failed to cast event's payload")
+	}
+
+	switch eventObj.EventName {
+	case "sched_process_exec":
+		err := sig.checkThread(&eventObj)
```

Comment:
Shouldn't you check if the event is from a process or a thread and then call the appropriate sig.checkXXX for it ? I'm concerned about this: in the checkProcess() you have:

```golang
	threadExist := false
	for tid := range processInfo.ThreadsIds {
		if tid == eventObj.HostThreadID {
			threadExist = true
			break
		}
	}
```

But this will only work if the task is a thread and not a process. What I'm saying is:

A process (or thread group leader) PID=1000 TID=1000 is executed.
- a proctree process is created (there is no thread 1000 on it)

A thread PID=1000 TID=1001 is executed:
- the proctree process entry for 1000 gets a thread 1001
- the proctree thread entry for 1001 is created

Reply:
In that case, the process tree is not consistent.
From the user pov, the process thread (tid == pid) is a thread. It should have a thread entry as well as a process entry.
For each TID in the system should be a matching thread in the tree, and for each PID in the system should be a matching process in the tree. This is the expected behavior when exposing a getter function for both.
I see that the problem in the implementation is that we coupled the `TaskInfo` into the `Process` struct, which doesn't make sense from user-mode pov. The process is the shared resources of all its threads, not a task. The current `TaskInfo` could be called also `ThreadInfo`, as the tasks in the kernel are thread info + process info (pointers to shared data like mm, etc.).
Anyways, do you want me to open an issue on it?

Reply:
https://github.com/aquasecurity/tracee/pull/3494